dvclive/metrics.py

```diff
@@ -100,10 +100,14 @@ def summary_path(self):
     def html_path(self):
         return self.dir + "_dvc_plots/index.html"
 
-    @property
-    def step(self):
+    def get_step(self) -> int:
         return self._step
 
+    def set_step(self, step: int):
+        if self._metrics:
+            self.next_step()
```

Comment:
I added this call mainly to preserve the behavior of a (probably corner case) scenario regarding the implicit `next_step` (discussed in #156 ). TBH the implicit `next_step `is not really documented and so this scenario would be difficult to reach.

Given the following code (originally included in tests):

```python
    steps = [0, 62, 1000]
    metrics = [0.9, 0.8, 0.7]

    for step, metric in zip(steps, metrics):
        dvclive.log("m", metric, step=step)
```

The latest summary (`latest.json`) generated would contain:

```json
{
    "step": 62,
    "m": 0.8
}
```

However, when removing `step` from `dvclive.log` and adding `dvclive.set_step`:

```python
    steps = [0, 62, 1000]
    metrics = [0.9, 0.8, 0.7]

    for step, metric in zip(steps, metrics):
        dvclive.set_step(step)
        dvclive.log("m", metric)
```


The latest summary (`latest.json`) generated would contain:

```json
{
    "step": 1000,
    "m": 0.8
}
```

Which is incorrect.

As I said, this is basically due to the `next_step` side effect in `dvclive.log`, which we could remove if we come to an angreement in #156 . 

The following code would work as intended and the `next_step` inside `set_step` will be just skipped (due to `if self._metrics: `) so the scenario you propose would behave as intended if the user calls `dvclive.next_step` manually:

```python
    steps = [0, 62, 1000]
    metrics = [0.9, 0.8, 0.7]

    for step, metric in zip(steps, metrics):
        dvclive.set_step(step)
        dvclive.log("m", metric)
        dvclive.next_step()
```

Reply:
Interesting. I think it makes sense to avoid asking user to call both `next_step` and `set_step` together. I also think the `{"step": 62, "m": 0.8}` summary would be more expected for users in case they forget to end their code with a call to `next_step`. Maybe we should be writing out the summary at each `log` call?

@pared Was there a reason for this original test to skip the summary for the last step?

Reply:
I discussed with @pared, and it seems like we may need to write out tsv and summary results at each `dvclive.log()` call so that:
1. `next_step` and `set_step` are consistent.
2. Only one of them is needed.
3. Behavior is consistent regardless of whether they are called at the beginning or end.

Reply:
Should that change be implemented before the one added in this P.R.?

Reply:
@dberenbaum main reason was to avoid too many I/O operations, but now it seems that current behaviour is standing in a way of correct behaviour of whole library, so I would not worry about it for now.

> Should that change be implemented before the one added in this P.R.?

I don't think it matters, as long as we make a release after both are incorporated to master.

Reply:
@dberenbaum @pared even if we consider the`summary` behavior fixed (by moving it to `log`), it doesn't feel like there is a way to fulfill the 3 points listed above when considering `html` and `checkpoints`.

Reply:
I think we should consider merging this P.R., if there are not any changes regarding the initial scope (remove `step` from `log` level,  and address the other changes raised during comments in separated issues/pr (i.e. #162 , #145 ). Does it makes sense?

Reply:
As long as it's consistent with the current `next_step`, let's merge.